bokeh/models/ranges.py

```diff
@@ -119,6 +142,16 @@ class FactorRange(Range):
     this categorical range.
     """)
 
+    bound_lower = Either(String, Int, help="""
+    The minimum value that the range is allowed to go to - typically used to prevent
+    the user from panning/zooming/etc away from the data.
+    """)
+
+    bound_upper = Either(String, Int, help="""
+    The max value that the range is allowed to go to - typically used to prevent
+    the user from panning/zooming/etc away from the data.
+    """)
```

Comment:
I'm not 100% certain this functionality is useful for categorical ranges, I think `upper` and `lower` imply an ordering, which is not the case here.